### PR TITLE
fix: SUPPORT-14018 - CUR extractor OOM fixes and improvements

### PR DESCRIPTION
**Pull request by ZdenekSrotyr**
# refactor: address PR review comments and inline manifest logging

## Summary
This PR addresses code quality improvements from PR review feedback, focusing on improving code readability and removing confusing terminology. The changes include refactoring the CUR 1.0 manifest retrieval logic with clearer helper methods, fixing nested `manifest.get()` calls, simplifying configuration validation, and removing the `_log_manifest_summary` helper method that was masking a simple list conversion operation.

**Key changes:**
- Refactored CUR 1.0 manifest retrieval with helper methods (`_is_manifest_file`, `_extract_period_from_path`, `_load_and_enrich_manifest`)
- Changed from wildcard CSV patterns to exact file paths from `reportKeys`
- Implemented period-based manifest selection (keeps latest by `LastModified` when multiple exist)
- Removed `_log_manifest_summary` helper and inlined logging directly in `retrieve_manifests`
- Removed all "deduplication" terminology, using "selected" instead
- Fixed nested `manifest.get()` calls by extracting to well-named variables
- Added whitespace handling for date configuration fields
- Simplified truthiness checks in configuration validation
- Moved `time` import to top of duckdb_client.py
- Updated docstrings for clarity

## Review & Testing Checklist for Human

- [ ] **CRITICAL: Verify component ID in `.github/workflows/push.yml`** - Currently set to `kds-team.ex-aws-cost-and-usage-reports-test` which appears to be a test configuration. This should likely be `kds-team.ex-aws-cost-and-usage-reports` for production deployment.

- [ ] **Test manifest retrieval logic** - The refactored `retrieve_manifests` method now uses exact file paths from `reportKeys` instead of wildcard patterns. Test with both CUR 1.0 folder structures:
  - `.../YYYYMMDD-YYYYMMDD/report-Manifest.json` (parent folder)
  - `.../YYYYMMDD-YYYYMMDD/timestampZ/report-Manifest.json` (grandparent folder with timestamp)

- [ ] **Verify period-based selection** - When multiple manifests exist for the same billing period (e.g., multiple timestamp folders), verify only the latest by `LastModified` is kept and no duplicate data is loaded

- [ ] **Test date configuration edge cases** - Test with whitespace in `max_date` and `min_date_since` fields (e.g., `" "`, `" now "`) to ensure stripping works correctly

- [ ] **Run integration tests** - Execute the component with actual CUR 1.0 and CUR 2.0 data to verify end-to-end functionality remains unchanged, especially the CSV pattern generation and data loading

### Notes
- All changes passed flake8 lint checks (pre-existing E501 line length warnings remain)
- The most significant change is in `cur_1_report_handler.py` where manifest retrieval was restructured - this is the highest risk area for functional changes
- The final commit addresses Martin's feedback about the "terrible coding pattern" where list conversion was hidden inside a logging method
- Large dependency updates in `uv.lock` are included but were not explicitly tested

**Requested by:** zdenek.srotyr@keboola.com (@ZdenekSrotyr)  
**Session:** https://app.devin.ai/sessions/cb68b455250641b18529e35360215f11

### CODE REVIEW

File: .github/workflows/push.yml [hunk 1]
```diff
@@ -4,7 +4,7 @@ on: [ push ]
 concurrency: ci-${{ github.ref }} # to avoid tag collisions in the ECR
 env:
   # repository variables:
-  KBC_DEVELOPERPORTAL_APP: "kds-team.ex-aws-cost-and-usage-reports" # replace with your component id
+  KBC_DEVELOPERPORTAL_APP: "kds-team.ex-aws-cost-and-usage-reports-test" # replace with your component id
```
Comment (soustruh) [inline .github/workflows/push.yml:7]: Plese fix the component id I guess? 🤞

File: src/aws_report_handlers/cur_1_report_handler.py [hunk 2]
```diff
@@ -26,46 +26,142 @@ class CUR1ReportHandler(BaseReportHandler):
...
+        - .../YYYYMMDD-YYYYMMDD/report-Manifest.json (parent folder)
+        - .../YYYYMMDD-YYYYMMDD/timestampZ/report-Manifest.json (grandparent folder)
+
+        Args:
+            key_parts: S3 key split by "/"
+
+        Returns:
+            Tuple of (period_folder_name, start_date) or (None, None) if not found
+        """
+        parent_folder_name = key_parts[-2]
```
Comment (soustruh) [inline src/aws_report_handlers/cur_1_report_handler.py:56]: Oh, the logic below is a bit messy… Instead of:
```py
        parent_folder_name = key_parts[-2]
        start_date, _ = self._try_parse_billing_period_from_folder_name(parent_folder_name)

        if not start_date and len(key_parts) >= 3:
            grandparent_folder_name = key_parts[-3]
            start_date, _ = self._try_parse_billing_period_from_folder_name(grandparent_folder_name)
            if start_date:
                return grandparent_folder_name, start_date

        if start_date:
            return parent_folder_name, start_date
```
I'd rather see
```py
        parent_folder_name = key_parts[-2]
        start_date, _ = self._try_parse_billing_period_from_folder_name(parent_folder_name)
        # this condition can go right here, the 59–63 block isn't executed if start_date evaluates
        # to True and if start_date produced inside the block, it's returned immediately anyway
        if start_date: 
            return parent_folder_name, start_date

        if len(key_parts) >= 3:  # "not start_date" is implicit here, no need to repeat it in the if clause
            grandparent_folder_name = key_parts[-3]
            start_date, _ = self._try_parse_billing_period_from_folder_name(grandparent_folder_name)
            if start_date:
                return grandparent_folder_name, start_date

```

File: src/aws_report_handlers/cur_1_report_handler.py [hunk 3]
```diff
@@ -26,46 +26,142 @@ class CUR1ReportHandler(BaseReportHandler):
...
+        manifest_content = self._read_s3_file_contents(s3_object["Key"])
+        manifest = json.loads(manifest_content)
+
+        manifest["last_modified"] = s3_object["LastModified"]
+        manifest["report_folder"] = s3_object["Key"].replace(f"/{manifest_pattern}", "")
+        manifest["period"] = period_folder
+        manifest["format_version"] = "1.0"
+
+        if "assemblyId" not in manifest:
+            manifest["assemblyId"] = manifest.get(
```
Comment (soustruh) [inline src/aws_report_handlers/cur_1_report_handler.py]: Is there a good reason for mixing underscore_case and camelCase in the manifest field names? Also, the `manifest.get()` as a part of a default value for another `manifest.get()` is kinda meh. 😐 I beg you, please assign the default value to a well-named variable (`default_value`, wink wink), this is quite difficult to read 👀 🙏

File: src/aws_report_handlers/cur_1_report_handler.py [hunk 4]
```diff
@@ -26,46 +26,142 @@ class CUR1ReportHandler(BaseReportHandler):
...
+
+        Args:
+            manifests_by_period: Dict mapping period folder to manifest
+            s3_objects: All S3 objects (for counting total manifests)
+            manifest_pattern: Manifest filename pattern (for counting)
+
+        Returns:
+            List of deduplicated manifests
+        """
+        manifests = list(manifests_by_period.values())
```
Comment (soustruh) [inline src/aws_report_handlers/cur_1_report_handler.py]: ❗ Calling `list()` to dict values doesn't deduplicate enything 👀 on the other hand, the manifests_by_period dict is already deduplicated anyway… 😀

File: src/aws_report_handlers/cur_1_report_handler.py [hunk 5]
```diff
@@ -26,46 +26,142 @@ class CUR1ReportHandler(BaseReportHandler):
...
+        CUR 1.0 manifests are located in report folders with date-based naming.
+        Supports both structures:
+        - .../YYYYMMDD-YYYYMMDD/report-Manifest.json
+        - .../YYYYMMDD-YYYYMMDD/timestampZ/report-Manifest.json
+
+        For nested structures with multiple timestamps per period, only the latest
+        manifest (by LastModified) is kept to avoid loading duplicate data.
         """
-        manifests = []
+        manifests_by_period = {}
```
Comment (soustruh) [inline src/aws_report_handlers/cur_1_report_handler.py:109]: In fact the manifests are already deduplicated by the simple fact you converted them to a dict… If you just want to cound original and resulting number of items, just put a counter somewhere and log the counts before calling return inside this method 🙏

File: src/configuration.py [hunk 6]
```diff
@@ -52,6 +52,10 @@ def end_datetime(self) -> datetime:
     @model_validator(mode="after")
     def validate_dates(self) -> "Configuration":
         """Validate date formats."""
+        # fix when old config had empty max_date
+        if self.max_date == "":
```
Comment (soustruh) [inline src/configuration.py]: I think I saw a space in the configuration when I checked it, do you strip the value somewhere or not? 🤷‍♂️

File: src/configuration.py [hunk 7]
```diff
@@ -52,6 +52,10 @@ def end_datetime(self) -> datetime:
...
+            self.max_date = "now"
+
         # Validate min_date_since
         if self.min_date_since is not None and self.min_date_since != "":
```
Comment (soustruh) [inline src/configuration.py]: Not part of the PR, I know… I'd still just do
```py
        if self.min_date_since:
```
🙃

File: src/duckdb_client.py [hunk 8]
```diff
@@ -1,135 +1,179 @@
...
-
-        logging.info(f"Initializing DuckDB connection with config: {config}")
-        conn = duckdb.connect(database=db_path, config=config)
-        return conn
+        Setup DuckDB connection with S3 credentials.
 
-    def setup_connection(self):
-        """Setup DuckDB connection with S3 credentials and performance
-        optimizations."""
+        Installs httpfs extension for S3 access and configures AWS credentials.
```
Comment (soustruh) [inline src/duckdb_client.py]: Installs httpfs extension? Where? ¯\\\_(ツ)_/¯

File: src/duckdb_client.py [hunk 9]
```diff
@@ -1,135 +1,179 @@
...
+        """
+        Apply S3 credentials and DuckDB settings to current connection.
 
-        logging.info("Setting up DuckDB connection...")
-        self.con = self._init_connection()
+        Used both for initial setup and after connection resets.
+        Settings based on Martin's Shopify component configuration.
+        """
+        self.con.execute("SET extension_directory='/tmp/duckdb_extensions';")
         self.con.execute("INSTALL httpfs;")
```
Comment (soustruh) [inline src/duckdb_client.py:47]: Oh, here 😁

File: src/duckdb_client.py [hunk 10]
```diff
@@ -1,135 +1,179 @@
...
+            logging.info("Closing DuckDB connection to release memory...")
+            self.con.close()
+            self.con = None
+            logging.info("DuckDB connection closed")
 
-        patterns_str = "', '".join(csv_patterns)
-        s3_count = sum(1 for p in csv_patterns if p.startswith("s3://"))
-        local_count = len(csv_patterns) - s3_count
+    @staticmethod
+    def _quote_ident(name: str) -> str:
```
Comment (soustruh) [inline src/duckdb_client.py:72]: Oh boy, DuckDB apparently really has no better way of handling identifiers with special characters. 🙀 I weep for the future of mankind 😁

File: src/duckdb_client.py [hunk 11]
```diff
@@ -1,135 +1,179 @@
...
+        automatically handle schema evolution.
+
+        Args:
+            csv_patterns: List of CSV file paths (S3 URIs or local paths)
+            final_columns: List of final column names in KBC format (with __)
+
+        Returns:
+            True if successful, False otherwise
+        """
+        import time
```
Comment (soustruh) [inline src/duckdb_client.py]: Please move the import to the top of the file 🙏

File: src/duckdb_client.py [hunk 12]
```diff
@@ -1,135 +1,179 @@
...
-                SELECT *
-                FROM read_csv_auto(['{patterns_str}'],
-                                   HEADER=TRUE,
-                                   ALL_VARCHAR=TRUE,
-                                   NULLSTR=['null', 'NULL', 'None'],
-                                   union_by_name=true,
-                                   filename=true);
+                CREATE OR REPLACE VIEW raw_unified AS
+                SELECT * EXCLUDE (filename)
+                FROM read_csv_auto([{file_list}],
```
Comment (soustruh) [inline src/duckdb_client.py:124]: Great discovery! 🌎 🥳

File: src/duckdb_client.py [hunk 13]
```diff
@@ -1,135 +1,179 @@
...
     def export_data_to_csv(self, output_path: str):
-        """Export data from DuckDB table to CSV file."""
-        self.con.execute(f"COPY {UNIFIED_REPORTS_VIEW} TO '{output_path}' (HEADER, DELIMITER ',');")
-        logging.info(f"Data exported to {output_path}")
+        """
+        Export data to CSV file using single COPY statement.
+
+        Exports entire unified view directly to output file in one pass.
+        """
+        import time
```
Comment (soustruh) [inline src/duckdb_client.py]: You see, importing time for the 2nd time 🥱

**Review by soustruh — CHANGES_REQUESTED** [review]
Gj, but please let's fix a couple of things. Let's meet in the office to discuss if you need any clarification, but I tried to explain everything in detail 🙏

File: src/aws_report_handlers/cur_1_report_handler.py (hunk 4 above)
Comment (soustruh) [reply src/aws_report_handlers/cur_1_report_handler.py]: Oh god, why do you focus on mentioning "deduplication" everywhere? There was no mention of it in the original code, the generated code probably introduced that during the process, now we are just removing what wasn't needed in the first place and the comments are all about deduplication deduplication, just forget deduplication, do/log the stuff you need to do and that's it.

Also, you take the manifests_by_period dict, pass it to the function called log_manifest_summary, then you call
```py
        manifests = list(manifests_by_period.values())
```
underhand and rely on the returned list of dict values – that is a seriously terrible coding pattern. Masking a simple oneline operation inside a method whose name doesn't suggest anything of this, that is just a what the 🐸 moment.

@ZdenekSrotyr please take care of this yourself, I am not particularly happy about having to explain this to a tin can. 😞 

Devin, don't you dare touch this again. 👎

**Review by soustruh — CHANGES_REQUESTED** [review]
Not happy, I don't write CRs for Devin, I take time to write them so another person also takes time, understands them and takes care of them appropriately 🙃 😢

**Comment by ZdenekSrotyr** [comment]
@soustruh how about that?

**Review by soustruh — APPROVED** [review]
OK, thanks for the cleanup 😇
